examples/example-clientlib-ktx-app/app/src/main/java/com/solanamobile/ktxclientsample/viewmodel/SampleViewModel.kt

```diff
@@ -3,41 +3,172 @@ package com.solanamobile.ktxclientsample.viewmodel
+
+                _state.value.copy(
+                    isLoading = false,
+                    solBalance = balance
+                ).updateViewState()
+            }
+        }
+    }
+
+    fun addFunds(sender: ActivityResultSender) {
         viewModelScope.launch {
-            walletAdapter.transact(sender) {
-                val result = authorize(Uri.parse("https://solana.com"), Uri.parse("favicon.ico"), "Solana")
+            val conn = persistanceUseCase.getWalletConnection()
+
+            val currentConn = walletAdapter.transact(sender) {
+                when (conn) {
+                   is NotConnected -> {
+                       val authed = authorize(solanaUri, iconUri, identityName, RpcCluster.Devnet)
+                       Connected(PublicKey(authed.publicKey), authed.authToken)
+                   }
+                   is Connected -> {
+                       val reauthed = reauthorize(solanaUri, iconUri, identityName, conn.authToken)
+                       Connected(PublicKey(reauthed.publicKey), reauthed.authToken)
+                   }
+                }
+            }
+
+            persistanceUseCase.persistConnection(currentConn.publickKey, currentConn.authToken)
+
+            _state.value.copy(
+                isLoading = true
+            ).updateViewState()
+
+            val tx = solanaRpcUseCase.requestAirdrop(currentConn.publickKey)
+            val confirmed = solanaRpcUseCase.awaitConfirmationAsync(tx).await()
+
+            if (confirmed) {
+                val balance = solanaRpcUseCase.getBalance(currentConn.publickKey)
+
+                _state.value.copy(
+                    isLoading = false,
+                    canTransact = true,
+                    solBalance = balance,
+                    userAddress = currentConn.publickKey.toBase58()
+                ).updateViewState()
+            } else {
+                _state.value.copy(
+                    isLoading = false,
+                    canTransact = true,
+                    solBalance = 0.0,
+                    userAddress = "Error airdropping"
+                ).updateViewState()
+            }
+        }
+    }
+
+    fun publishMemo(sender: ActivityResultSender, memoText: String) {
+        val conn = persistanceUseCase.getWalletConnection()
+
+        if (conn is Connected) {
+            _state.value.copy(
+                isLoading = true
+            ).updateViewState()
+
+            viewModelScope.launch {
+                val blockHash = solanaRpcUseCase.getLatestBlockHash()
+
+                val tx = Transaction()
+                tx.add(MemoProgram.writeUtf8(conn.publickKey, memoText))
+                tx.setRecentBlockHash(blockHash!!)
+                tx.feePayer = conn.publickKey
+
+                val bytes = tx.serialize(SerializeConfig(requireAllSignatures = false))
+
+                val result = walletAdapter.transact(sender) {
+                    reauthorize(solanaUri, iconUri, identityName, conn.authToken)
```

Comment:
The user won't be able to select this option until after they have authorized, so that's why we are assuming reauth. But like above, what's not handled is if the wallet tells us the token is invalid. Now I'm wondering too if I need to re-persist what is returned from reauthorize...